### ISSUE TITLE
Lime crashes on startup when compiling for Android with -debug

### ISSUE THREAD

**Issue by tbrosman** [openfl/lime#457]
I am not sure if this is a Lime bug or an hxcpp bug. I suspect Lime because I have reproduced the same behavior consistently regardless of the hxcpp version I'm using (I've tried some 3.1.x versions as well as 3.2.102) but it looks like a possible GC bug in hxcpp. The same code works in release mode, but I would like to be able to debug my android builds if possible :)

I've narrowed the behavior down to recent versions of lime (2.1.3 works without crashing, 2.3.3 crashes, 2.4.1 crashes) but don't know the exact version yet. Note that while it will crash consistently for some code, changing it slightly will alter the behavior: in some cases if I add a trace() statement to my code it'll crash later or not crash at all (I don't have a small example to post yet because of this). I'm hoping that someone can tell me whether or not this is Lime-related before I continue trying to narrow this down. Let me know if I should move this to the hxcpp issues instead or if there is any additional information I can provide to help narrow it down.

Example set of haxelib versions where I get this crash: openfl 3.0.6 + lime 2.4.1 + hxcpp 3.2.102 + flixel 3.3.8 compiling with haxe 3.1.3.

Backtrace:

```
I/DEBUG   (  177):     #00  pc 00a54d74  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::MarkObjectAlloc(hx::Object*, hx::MarkContext*)+32)
I/DEBUG   (  177):     #01  pc 000ca564  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (void hx::MarkMember<Dynamic>(Dynamic&, hx::MarkContext*)+48)
I/DEBUG   (  177):     #02  pc 000cba04  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (Array_obj<Dynamic>::__Mark(hx::MarkContext*)+92)
I/DEBUG   (  177):     #03  pc 00a54fc0  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::MarkObjectAlloc(hx::Object*, hx::MarkContext*)+620)
I/DEBUG   (  177):     #04  pc 000ca564  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (void hx::MarkMember<Dynamic>(Dynamic&, hx::MarkContext*)+48)
I/DEBUG   (  177):     #05  pc 00a82470  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::Hash<hx::TStringElement<Dynamic> >::HashMarker::operator()(hx::TStringElement<Dynamic>&)+136)
I/DEBUG   (  177):     #06  pc 00a785a8  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (void hx::Hash<hx::TStringElement<Dynamic> >::iterate<hx::Hash<hx::TStringElement<Dynamic> >::HashMarker>(hx::Hash<hx::TStringElement<Dynamic> >::HashMarker&)+68)
I/DEBUG   (  177):     #07  pc 00a6a670  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::Hash<hx::TStringElement<Dynamic> >::__Mark(hx::MarkContext*)+80)
I/DEBUG   (  177):     #08  pc 00a54fc0  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::MarkObjectAlloc(hx::Object*, hx::MarkContext*)+620)
I/DEBUG   (  177):     #09  pc 000ca564  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (void hx::MarkMember<Dynamic>(Dynamic&, hx::MarkContext*)+48)
I/DEBUG   (  177):     #10  pc 00a4a5b4  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::Anon_obj::__Mark(hx::MarkContext*)+44)
I/DEBUG   (  177):     #11  pc 00a54a7c  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::MarkContext::Process()+156)
I/DEBUG   (  177):     #12  pc 00a57b7c  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (GlobalAllocator::MarkAll(bool)+564)
I/DEBUG   (  177):     #13  pc 00a57d74  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (GlobalAllocator::Collect(bool, bool)+436)
```

Full debug output from logcat: https://gist.github.com/tbrosman/4c111ef2de793e29cb3f

**Comment by tbrosman** [comment]
After playing some version battleships I think I have a better idea of what's going on. There are two separate crashes: one when I change OpenFL versions and one when I change Lime versions (beyond the highest good set of OpenFL/Lime versions). I didn't bother to check the logcat output for the higher verisons the first time and assumed it was all the same issue.

lime 2.3.0 + openfl 2.1.8 - Working on Android in -debug builds
lime 2.3.0 + openfl 2.2.0 - Crashes in debug builds (garbage collector error above) but runs fine in release
lime 2.3.1 + openfl 2.1.8 - Doesn't compile
lime 2.3.2 + openfl 2.1.8 - Crashes in both debug and release; liblime-legacy bug below (note that I am not compiling with the legacy/lime-legacy flag)

Lime legacy crash backtrace:
```
I/DEBUG   (15702):     #00  pc 00000000  <unknown>
I/DEBUG   (15702):     #01  pc 00081ca8  /data/app-lib/com.vacuumTubeGames.myGame-1/liblime-legacy.so
I/DEBUG   (15702):     #02  pc 0002eb3c  /data/app-lib/com.vacuumTubeGames.myGame-1/liblime-legacy.so (Java_org_haxe_lime_Lime_onActivity+68)
```

Full logcat output: https://gist.github.com/tbrosman/c85811b1ae8e89db0b4d

Based on this the GC crash I started with looks like an OpenFL issue and the liblime-legacy crash either has something to do with my compile parameters or is a version mismatch with OpenFL. Should I file a separate bug in the openfl repo for the first issue?

**Comment by tbrosman** [comment]
My openfl/lime build battleships chart: https://gist.github.com/tbrosman/7eddf814e07ce5f1736c

At this point it's safe to say there are multiple issues across both OpenFL and Lime contributing to developer pain for anyone attempting to debug android builds. This seems like a good opportunity to track down these regressions and make future builds (of non-trivial projects) more resilient. In other words: you can learn from our explosions! :)

Again, if this is the wrong place for this please let me know.

**Comment by jgranick** [comment]
Unless you were using -Dlegacy, I noticed that an oversight caused the lime binary to not be copied in Lime 2.4.1 (perhaps before) causing clean Android builds to fail to start, but this was resolved in 2.4.2, I personally ran many samples on Android

I'll take a look at legacy too

**Comment by jgranick** [comment]
I wish it were simpler to unit test Android along with everything else :)

**Comment by tbrosman** [comment]
After I went back and tested more yesterday it appears the lime 2.3.0 + openfl 2.1.8 combination I listed actually does crash, but only when you start it from the OS (a '4' in my build matrix linked above).

The crash is similar to the GC crash from above, but is a SEGV_ACCERR instead of a SEGV_MAPERR. Backtrace:

```
I/DEBUG   (18046):     #00  pc 00044c70  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so
I/DEBUG   (18046):     #01  pc 00a15e38  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::MarkObjectAlloc(hx::Object*, hx::MarkContext*)+620)
I/DEBUG   (18046):     #02  pc 000c7d04  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (void hx::MarkMember<Dynamic>(Dynamic&, hx::MarkContext*)+48)
I/DEBUG   (18046):     #03  pc 000c91a4  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (Array_obj<Dynamic>::__Mark(hx::MarkContext*)+92)
I/DEBUG   (18046):     #04  pc 00a158f4  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (hx::MarkContext::Process()+156)
I/DEBUG   (18046):     #05  pc 00a189f4  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (GlobalAllocator::MarkAll(bool)+564)
I/DEBUG   (18046):     #06  pc 00a18bec  /data/app-lib/com.vacuumTubeGames.myGame-1/libApplicationMain.so (GlobalAllocator::Collect(bool, bool)+436)
```

Full logcat output: https://gist.github.com/tbrosman/bea1fd0366520cfc13b2

Not sure if this is related to the other GC crash or just more noise. Try exiting out of an app compiled with debug and then re-opening from the OS.

This is how I build and run normally:

```
haxelib run openfl build android -debug
haxelib run openfl run android -debug -Dfdb
```

**Comment by tbrosman** [comment]
By the way, what do you normally use as a test project? I tried reusing one of the Flixel demos for that purpose but it seemed like it would take a while to update the dependencies/calls to match anything recent. I tried making my own minimal project to repro the first GC crash, but got odd results depending on things like where trace was called and gave up. I'll give it another shot if you can't repro with what you're using.

**Comment by tbrosman** [comment]
@jgranick: My project randomly started working at a recent revision. I deleted things out of it until it broke, and the thing that broke it was deleting an import statement. The object wasn't even being allocated/used so I suspected dead code elimination. I turned it off in my application.xml and everything magically worked:

```xml
<haxeflag name="-dce" value="no" if="debug" />
```

I suspect the hxcpp GC behaves differently in debug builds from release builds. What lime version did you switch over to hxcpp? I'm guessing it lines up with where I have 5's in my original chart https://gist.github.com/tbrosman/7eddf814e07ce5f1736c

PS: Do you happen to know how to chain together conditions in if statements inside haxeflags? I Would like to have if="debug && android" but it didn't work when I tried it.

**Comment by jgranick** [comment]
Building C++ projects using DCE could bear more testing, sometimes, it seems it does not work fully with HXCPP

I wanted "a && b" support but that is invalid XML, you can do "and" using a space, similar to CSS:

```xml
<haxeflag name="-dce" value="no" if="android debug" />
```

**Comment by tbrosman** [comment]
Can't repro this anymore (fortunately). I have been consistently updating my openfl/lime/hxcpp packages over time and my theory is that the fixes to the GC in hxcpp 3.2.193 (https://github.com/HaxeFoundation/hxcpp/blob/master/Changes.md) fixed this issue for me.
